File: content/01-arrays-and-hashing/subarray_zero_sum.js
/**
 * Problem: Subarray with 0 Sum
 * GeeksforGeeks: https://practice.geeksforgeeks.org/problems/subarray-with-0-sum/0
 */

function hasZeroSumSubarray(arr) {
    /**
     * Time Complexity: O(N)
     * Space Complexity: O(N)
     */
    const prefixSumSet = new Set();
    let currSum = 0;
    
    for (let i = 0; i < arr.length; i++) {
        currSum += arr[i];
        if (currSum === 0 || prefixSumSet.has(currSum)) {
            return true;
        }
        prefixSumSet.add(currSum);
    }
    
    return false;
}

if (require.main === module) {
    const assert = require('assert');
    assert.strictEqual(hasZeroSumSubarray([4, 2, -3, 1, 6]), true);
    assert.strictEqual(hasZeroSumSubarray([4, 2, 0, 1, 6]), true);
    assert.strictEqual(hasZeroSumSubarray([1, 2, 3, 4]), false);
    console.log("JS: hasZeroSumSubarray passed tests!");
}

File: content/01-arrays-and-hashing/subarray_zero_sum.py
"""
Problem: Subarray with 0 Sum
GeeksforGeeks: https://practice.geeksforgeeks.org/problems/subarray-with-0-sum/0
"""

def has_zero_sum_subarray(arr: list[int]) -> bool:
    """
    Time Complexity: O(N)
    Space Complexity: O(N)
    """
    prefix_sum_set = set()
    curr_sum = 0
    
    for num in arr:
        curr_sum += num
        if curr_sum == 0 or curr_sum in prefix_sum_set:
            return True
        prefix_sum_set.add(curr_sum)
        
    return False

if __name__ == "__main__":
    assert has_zero_sum_subarray([4, 2, -3, 1, 6]) == True
    assert has_zero_sum_subarray([4, 2, 0, 1, 6]) == True
    assert has_zero_sum_subarray([1, 2, 3, 4]) == False
    print("Python: has_zero_sum_subarray passed tests!")
